## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :x: [more...](#8051d9e0-50f0-4bea-b6cd-e318e4a99c22) |
| mint-zoned.sh | :x: [more...](#c6ab393d-b51c-4c19-bf33-20945d098cb9) |
| mint-compress-encrypt-dist-erasure.sh | :x: [more...](#8d42f34e-b660-4f10-81c6-b9f8bb5bd3f6) |
<a name="c6ab393d-b51c-4c19-bf33-20945d098cb9"></a>
### 12139-accbf1a/mint-zoned.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-c3.minio.io:31370
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 0e8789f7f80d:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 2 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 42 seconds
(4/15) Running aws-sdk-ruby tests ... done in 4 seconds
(5/15) Running awscli tests ... done in 2 minutes and 17 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 1 minutes and 6 seconds
(8/15) Running minio-dotnet tests ... done in 54 seconds
(9/15) Running minio-go tests ... done in 2 minutes and 2 seconds
(10/15) Running minio-java tests ... done in 1 minutes and 51 seconds
(11/15) Running minio-js tests ... FAILED in 15 seconds
{
  "name": "minio-js",
  "function": "\"after all\" hook in \"functional tests\"",
  "duration": 5,
  "status": "FAIL",
  "error": "S3Error: The bucket you tried to delete is not empty at Object.parseError (node_modules/minio/dist/main/xml-parsers.js:79:11) at /mint/run/core/minio-js/node_modules/minio/dist/main/transformers.js:156:22 at DestroyableTransform._flush (node_modules/minio/dist/main/transformers.js:80:10) at DestroyableTransform.prefinish (node_modules/readable-stream/lib/_stream_transform.js:129:10) at prefinish (node_modules/readable-stream/lib/_stream_writable.js:611:14) at finishMaybe (node_modules/readable-stream/lib/_stream_writable.js:620:5) at endWritable (node_modules/readable-stream/lib/_stream_writable.js:643:3) at DestroyableTransform.Writable.end (node_modules/readable-stream/lib/_stream_writable.js:571:22) at IncomingMessage.onend (internal/streams/readable.js:684:10) at endReadableNT (internal/streams/readable.js:1327:12) at processTicksAndRejections (internal/process/task_queues.js:80:21)"
}
(11/15) Running minio-py tests ... FAILED in 2 minutes and 42 seconds
{
  "name": "minio-py:test_thread_safe",
  "status": "FAIL",
  "args": {
    "bucket_name": "minio-py-test-742a47f0-8b14-46ff-bfed-77d76555b57a",
    "object_name": "6f725891-6629-4099-934e-a8b29ac91226"
  },
  "message": "S3 operation failed; code: BucketNotEmpty, message: The bucket you tried to delete is not empty, resource: /minio-py-test-742a47f0-8b14-46ff-bfed-77d76555b57a, request_id: 16789F299BA4D9EC, host_id: c9bf51c4-4e97-44f1-abf6-4e80c9dd76b6, bucket_name: minio-py-test-742a47f0-8b14-46ff-bfed-77d76555b57a",
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 126, in _call_test\n    func(log_entry, *args, **kwargs)\n  File \"/mint/run/core/minio-py/tests.py\", line 1640, in test_thread_safe\n    _CLIENT.remove_bucket(bucket_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 676, in remove_bucket\n    self._execute(\"DELETE\", bucket_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 409, in _execute\n    preload_content=preload_content,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 385, in _url_open\n    raise response_error\nminio.error.S3Error: S3 operation failed; code: BucketNotEmpty, message: The bucket you tried to delete is not empty, resource: /minio-py-test-742a47f0-8b14-46ff-bfed-77d76555b57a, request_id: 16789F299BA4D9EC, host_id: c9bf51c4-4e97-44f1-abf6-4e80c9dd76b6, bucket_name: minio-py-test-742a47f0-8b14-46ff-bfed-77d76555b57a\n",
  "duration": 7312
}
(11/15) Running s3cmd tests ... done in 19 seconds
(12/15) Running s3select tests ... done in 6 seconds
(13/15) Running security tests ... done in 0 seconds

Executed 13 out of 15 tests successfully.
```
<a name="8051d9e0-50f0-4bea-b6cd-e318e4a99c22"></a>
### 12139-accbf1a/mint-gateway-s3.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev4.minio.io:31459
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp a762cba2aaa7:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 1 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 42 seconds
(4/15) Running aws-sdk-ruby tests ... done in 2 seconds
(5/15) Running awscli tests ... done in 2 minutes and 1 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 52 seconds
(8/15) Running minio-dotnet tests ... done in 38 seconds
(9/15) Running minio-go tests ... done in 54 seconds
(10/15) Running minio-java tests ... done in 45 seconds
(11/15) Running minio-js tests ... FAILED in 23 seconds
{
  "name": "minio-js",
  "function": "fPutObject(bucketName, objectName, filePath, metaData)",
  "args": "bucketName:minio-js-test-df4abc8d-fc9b-4082-b5d9-40f02c0c9223, objectName:datafile-65-MB, filePath:/tmp/datafile-65-MB",
  "duration": 855,
  "status": "FAIL",
  "error": "Error: read ECONNRESET at TCP.onStreamRead (internal/stream_base_commons.js:209:20)"
}
(11/15) Running minio-py tests ... done in 1 minutes and 23 seconds
(12/15) Running s3cmd tests ... done in 1 minutes and 5 seconds
(13/15) Running s3select tests ... done in 3 seconds
(14/15) Running security tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
<a name="8d42f34e-b660-4f10-81c6-b9f8bb5bd3f6"></a>
### 12139-accbf1a/mint-compress-encrypt-dist-erasure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-c3.minio.io:31171
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 63516d7a30b7:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 2 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 43 seconds
(4/15) Running aws-sdk-ruby tests ... done in 4 seconds
(5/15) Running awscli tests ... FAILED in 32 seconds
{
  "name": "awscli",
  "duration": 2749,
  "function": "aws --endpoint-url http://minio-c3.minio.io:31171 s3api copy-object --bucket awscli-mint-test-bucket-27494 --key datafile-1-kB-copy --copy-source awscli-mint-test-bucket-27494/datafile-1-kB\n",
  "status": "FAIL",
  "error": "Hash mismatch expected 084e1383b70fb0c51acc680fef370023, got ac57de7156d7fc25ac1a65f81fa3989b"
}
(5/15) Running healthcheck tests ... done in 0 seconds
(6/15) Running mc tests ... done in 48 seconds
(7/15) Running minio-dotnet tests ... done in 35 seconds
(8/15) Running minio-go tests ... FAILED in 2 minutes and 19 seconds
{
  "args": {},
  "duration": 381,
  "error": "At least one of the pre-conditions you specified did not hold",
  "function": "CopyObjectPart(destination, source)",
  "message": "CopyObjectPart call failed",
  "name": "minio-go: testUnencryptedToSSES3CopyObjectPart",
  "status": "FAIL"
}
(8/15) Running minio-java tests ... FAILED in 2 minutes and 28 seconds
{
  "name": "minio-java",
  "function": "copyObject()",
  "args": "[match etag]",
  "duration": 126,
  "status": "FAIL",
  "error": "error occurred\nErrorResponse(code = PreconditionFailed, message = At least one of the pre-conditions you specified did not hold, bucketName = minio-java-test-18gkfj3, objectName = minio-java-test-1rf4i6d-copy, resource = /minio-java-test-18gkfj3/minio-java-test-1rf4i6d-copy, requestId = 16789EEC0D809EAB, hostId = f48f5719-33bf-4ef4-9413-e638f6b2d136)\nrequest={method=PUT, url=http://minio-c3.minio.io:31171/minio-java-test-18gkfj3/minio-java-test-1rf4i6d-copy, headers=x-amz-copy-source-if-match: 71cff0a060f852067e443ad1e24ae26c-1\nx-amz-copy-source: /minio-java-test-2qbo4j2/minio-java-test-1rf4i6d\nHost: minio-c3.minio.io:31171\nAccept-Encoding: identity\nUser-Agent: MinIO (Linux; amd64) minio-java/8.0.3\nContent-MD5: 1B2M2Y8AsgTpgAmY7PhCfg==\nx-amz-content-sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855\nx-amz-date: 20210423T230002Z\nAuthorization: AWS4-HMAC-SHA256 Credential=*REDACTED*/20210423/us-east-1/s3/aws4_request, SignedHeaders=content-md5;host;x-amz-content-sha256;x-amz-copy-source;x-amz-copy-source-if-match;x-amz-date, Signature=*REDACTED*\n}\nresponse={code=412, headers=Accept-Ranges: bytes\nContent-Length: 418\nContent-Security-Policy: block-all-mixed-content\nContent-Type: application/xml\nETag: \"71cff0a060f852067e443ad1e24ae26c\"\nLast-Modified: Fri, 23 Apr 2021 23:00:02 GMT\nServer: MinIO\nVary: Origin\nX-Amz-Request-Id: 16789EEC0D809EAB\nX-Xss-Protection: 1; mode=block\nDate: Fri, 23 Apr 2021 23:00:02 GMT\n}\n >>> [io.minio.MinioClient.execute(MinioClient.java:775), io.minio.MinioClient.execute(MinioClient.java:563), io.minio.MinioClient.executePut(MinioClient.java:904), io.minio.MinioClient.copyObject(MinioClient.java:1232), FunctionalTest.testCopyObjectMatchETag(FunctionalTest.java:1850), FunctionalTest.copyObject(FunctionalTest.java:2016), FunctionalTest.runObjectTests(FunctionalTest.java:3757), FunctionalTest.runTests(FunctionalTest.java:3783), FunctionalTest.main(FunctionalTest.java:3927)]"
}
(8/15) Running minio-js tests ... done in 46 seconds
(9/15) Running minio-py tests ... done in 2 minutes and 35 seconds
(10/15) Running s3cmd tests ... FAILED in 5 seconds
{
  "name": "s3cmd",
  "duration": "2729",
  "function": "test_put_object_multipart",
  "status": "FAIL",
  "error": "WARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Too many failures. Giving up on '/mint/data/datafile-65-MB'\nERROR: \nUpload of '/mint/data/datafile-65-MB' part 1 failed. Use\n  /usr/local/bin/s3cmd abortmp s3://s3cmd-test-bucket-11373/s3cmd-test-object-2940 a8c1b1b6-284e-4bac-a8a7-5eec06d66f8e\nto abort the upload, or\n  /usr/local/bin/s3cmd --upload-id a8c1b1b6-284e-4bac-a8a7-5eec06d66f8e put ...\nto continue the upload.\nERROR: Upload of '/mint/data/datafile-65-MB' failed too many times (Last reason: )"
}
(10/15) Running s3select tests ... done in 6 seconds
(11/15) Running security tests ... done in 0 seconds

Executed 11 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

--- a/cmd/erasure-multipart.go
+++ b/cmd/erasure-multipart.go
@@ -301,11 +301,9 @@ func (er erasureObjects) newMultipartUpload(ctx context.Context, bucket string, 
 	partsMetadata := make([]FileInfo, len(onlineDisks))
 
 	fi := newFileInfo(pathJoin(bucket, object), dataDrives, parityDrives)
-	if opts.Versioned {
-		fi.VersionID = opts.VersionID
-		if fi.VersionID == "" {
-			fi.VersionID = mustGetUUID()
-		}
+	fi.VersionID = opts.VersionID
+	if opts.Versioned && fi.VersionID == "" {
+		fi.VersionID = mustGetUUID()
 	}
 	fi.DataDir = mustGetUUID()
 

--- a/cmd/erasure-object.go
+++ b/cmd/erasure-object.go
@@ -1433,6 +1433,7 @@ func (er erasureObjects) restoreTransitionedObject(ctx context.Context, bucket s
 	}
 
 	oi = actualfi.ToObjectInfo(bucket, object)
+	ropts := putRestoreOpts(bucket, object, opts.Transition.RestoreRequest, oi)
 	if len(oi.Parts) == 1 {
 		var rs *HTTPRangeSpec
 		gr, err := getTransitionedObjectReader(ctx, bucket, object, rs, http.Header{}, oi, opts)
@@ -1445,15 +1446,16 @@ func (er erasureObjects) restoreTransitionedObject(ctx context.Context, bucket s
 			return setRestoreHeaderFn(oi, toObjectErr(err, bucket, object))
 		}
 		pReader := NewPutObjReader(hashReader)
-		ropts := putRestoreOpts(bucket, object, opts.Transition.RestoreRequest, oi)
 		ropts.UserDefined[xhttp.AmzRestore] = completedRestoreObj(opts.Transition.RestoreExpiry).String()
 		_, err = er.PutObject(ctx, bucket, object, pReader, ropts)
 		return setRestoreHeaderFn(oi, toObjectErr(err, bucket, object))
 	}
-	uploadID, err := er.NewMultipartUpload(ctx, bucket, object, opts)
+
+	uploadID, err := er.NewMultipartUpload(ctx, bucket, object, ropts)
 	if err != nil {
 		return setRestoreHeaderFn(oi, err)
 	}
+
 	var uploadedParts []CompletePart
 	var rs *HTTPRangeSpec
 	// get reader from the warm backend - note that even in the case of encrypted objects, this stream is still encrypted.
@@ -1462,6 +1464,7 @@ func (er erasureObjects) restoreTransitionedObject(ctx context.Context, bucket s
 		return setRestoreHeaderFn(oi, err)
 	}
 	defer gr.Close()
+
 	// rehydrate the parts back on disk as per the original xl.meta prior to transition
 	for _, partInfo := range oi.Parts {
 		hr, err := hash.NewReader(gr, partInfo.Size, "", "", partInfo.Size)
@@ -1478,40 +1481,17 @@ func (er erasureObjects) restoreTransitionedObject(ctx context.Context, bucket s
 		})
 	}
 
-	uploadIDPath := er.getUploadIDDir(bucket, object, uploadID)
-	storageDisks := er.getDisks()
-	// Read metadata associated with the object from all disks.
-	partsMetadata, errs := readAllFileInfo(ctx, storageDisks, minioMetaMultipartBucket, uploadIDPath, "", false)
-
-	// get Quorum for this object
-	_, writeQuorum, err := objectQuorumFromMeta(ctx, partsMetadata, errs, er.defaultParityCount)
-	if err != nil {
-		return setRestoreHeaderFn(oi, toObjectErr(err, bucket, object))
-	}
-
-	reducedErr := reduceWriteQuorumErrs(ctx, errs, objectOpIgnoredErrs, writeQuorum)
-	if reducedErr == errErasureWriteQuorum {
-		return setRestoreHeaderFn(oi, toObjectErr(reducedErr, bucket, object))
-	}
-
-	_, modTime, dataDir := listOnlineDisks(storageDisks, partsMetadata, errs)
-
-	// Pick one from the first valid metadata.
-	fi, err := pickValidFileInfo(ctx, partsMetadata, modTime, dataDir, writeQuorum)
-	if err != nil {
-		return setRestoreHeaderFn(oi, err)
-	}
-	//validate parts created via multipart to transitioned object's parts info in xl.meta
 	partsMatch := true
-	if len(actualfi.Parts) != len(fi.Parts) {
-		partsMatch = false
-	}
-	if len(actualfi.Parts) == len(fi.Parts) {
-		for i, pi := range actualfi.Parts {
-			if fi.Parts[i].Size != pi.Size {
+	// validate parts created via multipart
+	if len(oi.Parts) == len(uploadedParts) {
+		for i, pi := range oi.Parts {
+			if uploadedParts[i].ETag != pi.ETag {
 				partsMatch = false
+				break
 			}
 		}
+	} else {
+		partsMatch = false
 	}
 
 	if !partsMatch {
@@ -1520,6 +1500,7 @@ func (er erasureObjects) restoreTransitionedObject(ctx context.Context, bucket s
 
 	_, err = er.CompleteMultipartUpload(ctx, bucket, object, uploadID, uploadedParts, opts)
 	if err != nil {
+		uploadIDPath := er.getUploadIDDir(bucket, object, uploadID)
 		return setRestoreHeaderFn(oi, toObjectErr(err, minioMetaMultipartBucket, uploadIDPath))
 	}
 	return setRestoreHeaderFn(oi, nil)

--- a/cmd/object-api-interface.go
+++ b/cmd/object-api-interface.go
@@ -52,13 +52,13 @@ type ObjectOptions struct {
 	DeleteMarkerReplicationStatus string                 // Is only set in DELETE operations
 	VersionPurgeStatus            VersionPurgeStatusType // Is only set in DELETE operations for delete marker version to be permanently deleted.
 	Transition                    TransitionOptions
-	NoLock                        bool                                                  // indicates to lower layers if the caller is expecting to hold locks.
-	ProxyRequest                  bool                                                  // only set for GET/HEAD in active-active replication scenario
-	ProxyHeaderSet                bool                                                  // only set for GET/HEAD in active-active replication scenario
-	ParentIsObject                func(ctx context.Context, bucket, parent string) bool // Used to verify if parent is an object.
 
-	// Use the maximum parity (N/2), used when
-	// saving server configuration files
+	NoLock         bool                                                  // indicates to lower layers if the caller is expecting to hold locks.
+	ProxyRequest   bool                                                  // only set for GET/HEAD in active-active replication scenario
+	ProxyHeaderSet bool                                                  // only set for GET/HEAD in active-active replication scenario
+	ParentIsObject func(ctx context.Context, bucket, parent string) bool // Used to verify if parent is an object.
+
+	// Use the maximum parity (N/2), used when saving server configuration files
 	MaxParity bool
 }
 

